src/erlcloud_ddb2.erl

```diff
@@ -2577,6 +2579,28 @@ list_tables(Opts, Config) ->
     out(Return, fun(Json, UOpts) -> undynamize_record(list_tables_record(), Json, UOpts) end, 
         DdbOpts, #ddb2_list_tables.table_names, {ok, []}).
 
+
+list_all_tables() ->
+    list_all_tables(default_config()).
+
+-spec list_all_tables(aws_config()) -> {ok, [table_name()]} | {error, any()}.
+list_all_tables(AWSCfg) ->
+    do_list_all_tables(undefined, AWSCfg, []).
+
+do_list_all_tables(LastTable, AWSCfg, Result) ->
+    Options = [{exclusive_start_table_name, LastTable}, {out, record}],
+    case erlcloud_ddb2:list_tables(Options, AWSCfg) of
+        {ok, #ddb2_list_tables{table_names = TableNames, last_evaluated_table_name = LastTableName}} ->
+            NewResult = TableNames ++ Result,
```

Comment:
This is probably a good use case for [flatreverse](https://github.com/erlcloud/erlcloud/blob/895fb3b3a0891ae21b56fd8c8029c8dd7a7dc611/src/erlcloud_ddb_util.erl#L505-L507) in [erlcloud_ddb_util](https://github.com/erlcloud/erlcloud/blob/master/src/erlcloud_ddb_util.erl).

Reply:
Sorry, I can't understand your mean. Result and TableNames are binary list, I think Result ++ TableNames could keep the order but why flatreverse is better than ++?

Thank you.

Reply:
The main reason is that flatreverse (see how it's used with [q_all](https://github.com/erlcloud/erlcloud/blob/895fb3b3a0891ae21b56fd8c8029c8dd7a7dc611/src/erlcloud_ddb_util.erl#L338) and [scan_all](https://github.com/erlcloud/erlcloud/blob/895fb3b3a0891ae21b56fd8c8029c8dd7a7dc611/src/erlcloud_ddb_util.erl#L401)) actually returns the tables in the order that they are returned by DynamoDB API, where as the above append strategy, while pretty efficient, appends the lists backwards (the last set of table names is at the beginning of the list). Flatreverse is as efficient as the above function but the order is symmetric to an in-order traversal of the table list from DynamoDB.

Reply:
I known the order problem, so I will use Result ++ TableNames to keep the order, but I don't agree with "efficient", in fact, flatreverse can't be used in here:

```
Tables = [<<"Tab1">>, <<"Tab2">>] ,  
Result = [<<"Tab3">>, <<"Tab4">>] ,  
erlcloud_ddb_util:flatreverse([Tables | Result]).  
```

output:

```
** exception error: bad argument  
     in operator  ++/2  
        called as <<"Tab3">> ++ [<<"Tab1">>,<<"Tab2">>]  
...  
```

English is not my first language, please ignore my English mistakes.  : )
Thank you.

Reply:
You can indeed switch the order using `Result ++ TableNames`, but this is comparatively inefficient because every element in `Result` ends up getting copied at every step of the function. This probably isn't terrible in most cases, but it's generally something to avoid, because it's the accumulated list and it can grow very large. It is best to append the small list into the larger one that is being accumulated.

I think you have a minor issue with your use above. Results should be accumulated as a list of lists, and then "reverse-flattened" at the end:

```erlang
T1 = [<<"table1">>, <<"table2">>],
T2 = [<<"table3">>, <<"table4">>],
T3 = [<<"table5">>, <<"table6">>],
erlcloud_ddb_util:flatreverse([T3, T2, T1]).
% [<<"table1">>,<<"table2">>,<<"table3">>,<<"table4">>,
%  <<"table5">>,<<"table6">>]
```

So, in your `list_tables_all` function, you would need to build the results as `[Tables | Result]` as you have it, but note that in the initial case, Result is `[]`, so you get `[[<<"Tab3">>, <<"Tab4">>]|[]]`, which is `[[<<"Tab3">>, <<"Tab4">>]]`. As you build from there, you get the backwards list of lists of table names that you then `flatreverse` at the end; this uses much less redundant list copying. The problem in your example above is just that you haven't expressed the `Result` variable correctly. If you implement the function as it is for `scan_all`,  it will just work in practice.

Reply:
Thanks for you tell me the correct way to use flatreverse.

I agree :
`It is best to append the small list into the larger one that is being accumulated.`, 
many Erlang books talk about this.

Normally, most user's tables less than one hundred or less than one thousand in a region, this is a simple test:

```erlang
flatreverse(Result, 0) ->
    Result;
flatreverse(Result, Max) ->
    List = erlcloud_ddb_util:flatreverse([[1,2,3,4,5,6,7,8,9,0], Result]),
    flatreverse(List, Max-1).


plus(Result, 0) ->
    Result;
plus(Result, Max) ->
    plus(Result ++ [1,2,3,4,5,6,7,8,9,0], Max-1).

tc(Max) ->
    {Time1, L1} = timer:tc(?MODULE, flatreverse, [[], Max]),
    {Time2, L2} = timer:tc(?MODULE, plus, [[], Max]),
    io:format("~p, ~p, ~p, ~p~n", [length(L1), length(L2), Time1, Time2]).
```  

output:

```
1> test:tc(1).
10, 10, 2, 0
ok
2> test:tc(50).
500, 500, 3035, 56
ok
3> test:tc(100).
1000, 1000, 372, 259
ok
4> test:tc(500).
5000, 5000, 9181, 9481
ok
5> test:tc(1000).
10000, 10000, 31010, 21884
ok
6> test:tc(5000).
50000, 50000, 502264, 448474
```

So I think `This probably isn't terrible in most cases` is enough, ++ and flatreverse both fine but ++ is more simple in this case.

Finally, I would like accept your suggestion and use flatreverse because the code lib should keep the same style and more safe than ++.

Thanks again. You are a nice man and good reviewer.